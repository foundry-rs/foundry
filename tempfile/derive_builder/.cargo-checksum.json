{"files":{"CHANGELOG.md":"48c03e5308887beede4a33a836dabd78fea7c493cf1bda67b6cb7b7ba2b12297","Cargo.lock":"db030f18562271de776d09d65806873bc2b88b81a8a5f651577105ed24921e75","Cargo.toml":"cb5f6811ad40ee5b7f6846989234e3566a2ba0c934ddb1b38f6b80510d1ca49c","LICENSE-APACHE":"c6596eb7be8581c18be736c846fb9173b69eccf6ef94c5135893ec56bd92ba08","LICENSE-MIT":"8c9612877aacfa1b42f5e80b679e6c3c93a7ba2bb99dfaf6e87e200dca4b4e6c","README.md":"c7e12ee517ff85c03176f91521f00e1576686e704dd8f985ae54d9565ffe7de6","examples/channel.rs":"f8d80bd91719773a19b65e53196559fcef8e6da5b7c74be90eeadd0117f9c972","examples/custom_constructor.rs":"72fc36d95dae0d3c3d8a69c490f63e23cd04f852ff93340f73ea4ca618b2fd74","examples/custom_defaults.rs":"f062b63f03c06a2cd95a343ed72154fe2331b05f6a803fd7e944484d4910f507","examples/custom_error.rs":"e2f0d0227d937db5cefc64046a77e2d4290fa28ec947d30c7cc384d7ae6e8cd6","examples/custom_error_generic.rs":"f2e8985866713d235b937a7cd35cf2bf7d2d6831fde154be42424a6b4a58c219","examples/deny_missing_docs.rs":"017346be1c4d2fa586a7aff6ac14760cad58c4735b4cd7f4eb75e2f32764e1da","examples/doc_example.rs":"ec43cacb7e7bea8199bc053da612a2846e0294e6255ec444c71934238207cbfd","examples/readme_example.rs":"d350a7c761858e3f03fbb82f6e58ba1bd80ab0856a61642048d884bc3e80f943","examples/validation.rs":"223765e94e140db946dc59317ae45bf16f86e427e27806b1617eab05025d22bd","src/error.rs":"5ea269b7df934a2acbc2afbc86730032b1e3a12f90d2f639bd4ed83184e93b86","src/lib.rs":"5e9e3ecd781e0e1baf88f76303bf35e72381835cabd12cea7139f24769aa2d77","tests/bounds_generation.rs":"ec3f5c3c696a5ea3bcc8f4352c66bf8a02a21618b17f89172af6daf06fde72d1","tests/build_fn.rs":"d8dceb0a6424dea1e5ac9b3ea5e907406b59b7806bd8c3ad857819f0828ec90f","tests/builder_field_custom.rs":"689780c615b3966a4c0e799a617d44e27145c0be30d88cd0784ae73eecf46304","tests/builder_name.rs":"d80a8d97096065e9920d94054e0310d7cf8a7096fc497c3060c853d63974b95e","tests/compile-fail/build_fn_error.rs":"9ee98833f2c257f3ae760a114d742002ab93e546ed7e136181da04fd88b0e93f","tests/compile-fail/build_fn_error.stderr":"2af5788ec66abdae2acfe5124f3959472f0150e1c97ec335fadf006af841c048","tests/compile-fail/builder_field_attr.rs":"30891d59916f6b0d724d9eae6f9ed7cf2c6a854b8ab365ac1aff3775e5b800f9","tests/compile-fail/builder_field_attr.stderr":"4c583063185e27ca97d9d16aa166cf03a4604a61e6010bebb3d0888d7d28f8c2","tests/compile-fail/builder_field_custom.rs":"d1bda6379f4590ddd123e71aade0871fd80f7a3bb10ce388a593a6146cda71a6","tests/compile-fail/builder_field_custom.stderr":"cdadd050f0da58da8a70152c8a0d22b2d736a37ae69eb72a51d194cb5a557898","tests/compile-fail/builder_setter_attr.rs":"76da57b36a59932315f591657dd0138839092e868d85a01f3327b064d6afa39d","tests/compile-fail/builder_setter_attr.stderr":"e9c1d9dc4697bcf08c8cb1335e2c987f19797598697a7ad0fb9798c5e3c1ec56","tests/compile-fail/crate_root.rs":"0f888cebd5c3f142f219671d57a51c7ce0865f6aeaebb5b7495837fe0be839e7","tests/compile-fail/crate_root.stderr":"b59e6a41b06f074598d2e221bd991085374ca3de69dd27c2fd1e90eb6569e173","tests/compile-fail/custom_error_generic_missing_bound.rs":"1fc990ab299b529ec7ef682721f5b94e3dac3502ea399d75b3a683fa60475ecb","tests/compile-fail/custom_error_generic_missing_bound.stderr":"7485b29ccfc04f48557f429d2de8b029bd3e9916dca804f8087ff15fb89e4717","tests/compile-fail/custom_error_no_from.rs":"342cf8bf61475001df6b2749c62570a54ca268042579cfe51a16656bee5fc184","tests/compile-fail/custom_error_no_from.stderr":"3f9b4a6bef48d8d41da7723b076ec0ef382d9fe088619ce81ea7c55117ce77be","tests/compile-fail/deny_empty_default.rs":"a32d134f0c5de1c25d6e32ce80b596a3a0235ffef29c2110af1fde3483989853","tests/compile-fail/deny_empty_default.stderr":"a59bee15b7828cc6411a2bbe3cdc227fa1c8af3ed3706696af0def854e6eb565","tests/compile-fail/private_build_fn.rs":"f07dc93e881366ade4ba321a1214b6292a79db8149e2994e934882a275b62f00","tests/compile-fail/private_build_fn.stderr":"4e755df8d1d49e39ffcebaa2b435fc8e98e298307b6a2a667b80487b7d43cafe","tests/compile-fail/private_builder.rs":"305011243ae60ecb8d66e6a85cd561e0494be2cc0c352f4c20b6a8440cccecd7","tests/compile-fail/private_builder.stderr":"9ad769774f9598b802d49504b203f5b7f1d6e57458859753f417d6a09f965d0c","tests/compile-fail/private_fields.rs":"698046752b8c70947b20b9efc4af1add696519b270b6829d21ffbe529de96818","tests/compile-fail/private_fields.stderr":"783d7fa892cf84c1bcb0b1f9e2bbe3abb63aa538d6c8708569a64682eee51538","tests/compile-fail/rename_setter_struct_level.rs":"1abd1863379230ec0e19b78ffcff36c45dc2ad51935e91b78f5b9cc76c880c3b","tests/compile-fail/rename_setter_struct_level.stderr":"e551ba5115df07f86f4b1c6eb89ac90dca1dbc8817fdca3b31fa245a9c9b98e9","tests/compile-fail/vis_conflict.rs":"4e3cb313210c329814509ccc88f0e0ef322728bfdbfa7930a6eea3742e6dd9d9","tests/compile-fail/vis_conflict.stderr":"8d515a02fa1de8a7fc0e9216d83409dfc9e59297aa1501868b38839a41f04092","tests/compiletests.rs":"c137b64ccade2328f210e987ddb48d2d1f3342b6566b9e4a55ff21aa336c2ca5","tests/custom_constructor.rs":"d934ee1e3ab68930a9cace0620eeacdbd38fe716088437eaedad553772de06ed","tests/custom_default.rs":"31f7db3517e8ce3a92a3297ac87d03222d2dc3eac956a4ed32a62ea2a4c9ff9c","tests/derive_trait.rs":"d930ab3f477f2d492001cd80d9d662351e0b95fcc29ab297e7dff0192e629977","tests/forward_allow_attr.rs":"cd18e603142c31eef3162397e4819feb6c350a96836487b8ddac6c80c3f21e43","tests/forward_serde_attrs.rs":"de011806eadc79d112f76a7c43e0b34d783927a0747f278eef7511533349718f","tests/generic_structs.rs":"350faaba301920b387c4e173345fe75b0c8983373184bdc84e513ff251d49651","tests/generic_with_default.rs":"209a154ef007bdc727fef3c5596f1ac8845a73d4b4ec22387da4028f0ab49ed6","tests/ignore/no_std.rs":"f09678cec7134baf5f8ecc569e2d2feb462a22f1fc3b34614030e3611591164d","tests/lifetime.rs":"ec4960fc5e24759fe8f4706f08fedf05a46572ff6c026f4b946717f8a64056a6","tests/run-pass/attributes.rs":"9b01b0041f975280a1ad0b261827b94a13c8839cf65569af1e2a38c39905c9d0","tests/run-pass/crate_alias.rs":"16a5783e0f26708f90212410d96154ff35d80c67545bb72c6630089cd6fe29c4","tests/run-pass/custom_error_default.rs":"1e816a4caaf38fab3dd1b62893e9258fc4bbbc75427019c5289bd63eca437f28","tests/run-pass/custom_types.rs":"323eb5f5f90476cf3e08f931a90679fde6fd3ede631b0a8c6944b1333ab962a4","tests/run-pass/default_typeparams.rs":"b56035b70671893100261353189c7cc2ae9aaa2066c52c6b32767b0a6356b9df","tests/run-pass/empty_struct.rs":"26991529dfb588fd98e4b2cfbb2012b8209ef62cd830492778d50239d7279442","tests/run-pass/multiple_derives.rs":"a3966a22bfdeb3a004f807fce70b273ae63b946d6c4fdc70089323139541b64e","tests/setter_custom.rs":"7584efdc2d9ffa68a8e044eac838f36c0a8c0c93826bbba1a00a5399dc098be8","tests/setter_extend.rs":"f098528ea4953b844beb0b76898d15591fcee30497ea517d2e5592e5912c4cfb","tests/setter_into.rs":"8c5dca939b78b20335afea37e4d83ca9ba8ed373a42ed458316fdb8eaada73d1","tests/setter_name.rs":"03083c34f40effbff8b2134568464e2736a8fc5a1fc1200af861fb1efae7890b","tests/setter_pattern.rs":"2819e6aa79f9e079982e1ada44a62949665a84848cf2f4e02ca51e66b85a89cf","tests/setter_prefix.rs":"59b261f3b1d2f475afd60ea114a05015a425d71173c7b3e3542b6d648296bb76","tests/setter_strip_option.rs":"6d7e833f9132563c52e5a15c80cf0e88a90a191828bac2756d9019eb05511d24","tests/setter_visibility.rs":"d4220ec8857852f90c6c5aadf2ec2629e80942ff837cac9bd7a58c4d034cf4b6","tests/skip-setter.rs":"51ca8a2c105f1748322ae5a5a8a2089476410c1e70d9a2d617c3f00502b3c628","tests/try_setter.rs":"8dc1ee166cf27ad4e993b38b2a05a538efcc4c0dc635943d13b0172451606181","tests/validation.rs":"5703541a677dfa9cb6290bc134e4905611bb10bdce631b27bcfdfa8a2cdffe6c"},"package":"507dfb09ea8b7fa618fcf76e953f4f5e192547945816d5358edffe39f6f94947"}