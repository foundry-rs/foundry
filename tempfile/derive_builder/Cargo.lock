# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "darling"
version = "0.20.10"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6f63b86c8a8826a49b8c21f08a2d07338eec8d900540f8630dc76284be802989"
dependencies = [
 "darling_core",
 "darling_macro",
]

[[package]]
name = "darling_core"
version = "0.20.10"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "95133861a8032aaea082871032f5815eb9e98cef03fa916ab4500513994df9e5"
dependencies = [
 "fnv",
 "ident_case",
 "proc-macro2",
 "quote",
 "strsim",
 "syn 2.0.48",
]

[[package]]
name = "darling_macro"
version = "0.20.10"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d336a2a514f6ccccaa3e09b02d41d35330c07ddf03a62165fcec10bb561c7806"
dependencies = [
 "darling_core",
 "quote",
 "syn 2.0.48",
]

[[package]]
name = "derive_builder"
version = "0.20.2"
dependencies = [
 "derive_builder_macro",
 "rustversion",
 "serde",
 "serde_json",
 "trybuild",
]

[[package]]
name = "derive_builder_core"
version = "0.20.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "2d5bcf7b024d6835cfb3d473887cd966994907effbe9227e8c8219824d06c4e8"
dependencies = [
 "darling",
 "proc-macro2",
 "quote",
 "syn 2.0.48",
]

[[package]]
name = "derive_builder_macro"
version = "0.20.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ab63b0e2bf4d5928aff72e83a7dace85d7bba5fe12dcc3c5a572d78caffd3f3c"
dependencies = [
 "derive_builder_core",
 "syn 2.0.48",
]

[[package]]
name = "fnv"
version = "1.0.7"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3f9eec918d3f24069decb9af1554cad7c880e2da24a9afd88aca000531ab82c1"

[[package]]
name = "glob"
version = "0.3.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9b919933a397b79c37e33b77bb2aa3dc8eb6e165ad809e58ff75bc7db2e34574"

[[package]]
name = "ident_case"
version = "1.0.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b9e0384b61958566e926dc50660321d12159025e767c18e043daf26b70104c39"

[[package]]
name = "itoa"
version = "1.0.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4217ad341ebadf8d8e724e264f13e593e0648f5b3e94b3896a5df283be015ecc"

[[package]]
name = "once_cell"
version = "1.15.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e82dad04139b71a90c080c8463fe0dc7902db5192d939bd0950f074d014339e1"

[[package]]
name = "proc-macro2"
version = "1.0.78"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e2422ad645d89c99f8f3e6b88a9fdeca7fabeac836b1002371c4367c8f984aae"
dependencies = [
 "unicode-ident",
]

[[package]]
name = "quote"
version = "1.0.35"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "291ec9ab5efd934aaf503a6466c5d5251535d108ee747472c3977cc5acc868ef"
dependencies = [
 "proc-macro2",
]

[[package]]
name = "rustversion"
version = "1.0.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "97477e48b4cf8603ad5f7aaf897467cf42ab4218a38ef76fb14c2d6773a6d6a8"

[[package]]
name = "ryu"
version = "1.0.11"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4501abdff3ae82a1c1b477a17252eb69cee9e66eb915c1abaa4f44d873df9f09"

[[package]]
name = "serde"
version = "1.0.147"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d193d69bae983fc11a79df82342761dfbf28a99fc8d203dca4c3c1b590948965"
dependencies = [
 "serde_derive",
]

[[package]]
name = "serde_derive"
version = "1.0.147"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4f1d362ca8fc9c3e3a7484440752472d68a6caa98f1ab81d99b5dfe517cec852"
dependencies = [
 "proc-macro2",
 "quote",
 "syn 1.0.103",
]

[[package]]
name = "serde_json"
version = "1.0.87"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6ce777b7b150d76b9cf60d28b55f5847135a003f7d7350c6be7a773508ce7d45"
dependencies = [
 "itoa",
 "ryu",
 "serde",
]

[[package]]
name = "strsim"
version = "0.11.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "7da8b5736845d9f2fcb837ea5d9e2628564b3b043a70948a3f0b778838c5fb4f"

[[package]]
name = "syn"
version = "1.0.103"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a864042229133ada95abf3b54fdc62ef5ccabe9515b64717bcb9a1919e59445d"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "syn"
version = "2.0.48"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0f3531638e407dfc0814761abb7c00a5b54992b849452a0646b7f65c9f770f3f"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "termcolor"
version = "1.1.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "bab24d30b911b2376f3a13cc2cd443142f0c81dda04c118693e35b3835757755"
dependencies = [
 "winapi-util",
]

[[package]]
name = "toml"
version = "0.5.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8d82e1a7758622a465f8cee077614c73484dac5b836c02ff6a40d5d1010324d7"
dependencies = [
 "serde",
]

[[package]]
name = "trybuild"
version = "1.0.71"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ea496675d71016e9bc76aa42d87f16aefd95447cc5818e671e12b2d7e269075d"
dependencies = [
 "glob",
 "once_cell",
 "serde",
 "serde_derive",
 "serde_json",
 "termcolor",
 "toml",
]

[[package]]
name = "unicode-ident"
version = "1.0.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6ceab39d59e4c9499d4e5a8ee0e2735b891bb7308ac83dfb4e80cad195c9f6f3"

[[package]]
name = "winapi"
version = "0.3.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5c839a674fcd7a98952e593242ea400abe93992746761e38641405d28b00f419"
dependencies = [
 "winapi-i686-pc-windows-gnu",
 "winapi-x86_64-pc-windows-gnu",
]

[[package]]
name = "winapi-i686-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ac3b87c63620426dd9b991e5ce0329eff545bccbbb34f3be09ff6fb6ab51b7b6"

[[package]]
name = "winapi-util"
version = "0.1.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "70ec6ce85bb158151cae5e5c87f95a8e97d2c0c4b001223f33a334e3ce5de178"
dependencies = [
 "winapi",
]

[[package]]
name = "winapi-x86_64-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "712e227841d057c1ee1cd2fb22fa7e5a5461ae8e48fa2ca79ec42cfc1931183f"
