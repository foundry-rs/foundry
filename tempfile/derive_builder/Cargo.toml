# THIS FILE IS AUTOMATICALLY GENERATED BY CARGO
#
# When uploading crates to the registry Cargo will automatically
# "normalize" Cargo.toml files for maximal compatibility
# with all versions of Cargo and also rewrite `path` dependencies
# to registry (e.g., crates.io) dependencies.
#
# If you are reading this file be aware that the original Cargo.toml
# will likely look very different (and much more reasonable).
# See Cargo.toml.orig for the original contents.

[package]
edition = "2018"
name = "derive_builder"
version = "0.20.2"
authors = [
    "Colin Kiegel <kiegel@gmx.de>",
    "Pascal Hertleif <killercup@gmail.com>",
    "Jan-Erik Rediger <janerik@fnordig.de>",
    "Ted Driggs <ted.driggs@outlook.com>",
]
build = false
autobins = false
autoexamples = false
autotests = false
autobenches = false
description = "Rust macro to automatically implement the builder pattern for arbitrary structs."
documentation = "https://docs.rs/derive_builder/0.20.2"
readme = "README.md"
keywords = [
    "derive",
    "macro",
    "builder",
    "setter",
    "struct",
]
categories = [
    "development-tools",
    "rust-patterns",
]
license = "MIT OR Apache-2.0"
repository = "https://github.com/colin-kiegel/rust-derive-builder"

[lib]
name = "derive_builder"
path = "src/lib.rs"

[[example]]
name = "channel"
path = "examples/channel.rs"

[[example]]
name = "custom_constructor"
path = "examples/custom_constructor.rs"

[[example]]
name = "custom_defaults"
path = "examples/custom_defaults.rs"

[[example]]
name = "custom_error"
path = "examples/custom_error.rs"

[[example]]
name = "custom_error_generic"
path = "examples/custom_error_generic.rs"

[[example]]
name = "deny_missing_docs"
path = "examples/deny_missing_docs.rs"

[[example]]
name = "doc_example"
path = "examples/doc_example.rs"

[[example]]
name = "readme_example"
path = "examples/readme_example.rs"

[[example]]
name = "validation"
path = "examples/validation.rs"

[[test]]
name = "bounds_generation"
path = "tests/bounds_generation.rs"

[[test]]
name = "build_fn"
path = "tests/build_fn.rs"

[[test]]
name = "builder_field_custom"
path = "tests/builder_field_custom.rs"

[[test]]
name = "builder_name"
path = "tests/builder_name.rs"

[[test]]
name = "compiletests"
path = "tests/compiletests.rs"

[[test]]
name = "custom_constructor"
path = "tests/custom_constructor.rs"

[[test]]
name = "custom_default"
path = "tests/custom_default.rs"

[[test]]
name = "derive_trait"
path = "tests/derive_trait.rs"

[[test]]
name = "forward_allow_attr"
path = "tests/forward_allow_attr.rs"

[[test]]
name = "forward_serde_attrs"
path = "tests/forward_serde_attrs.rs"

[[test]]
name = "generic_structs"
path = "tests/generic_structs.rs"

[[test]]
name = "generic_with_default"
path = "tests/generic_with_default.rs"

[[test]]
name = "lifetime"
path = "tests/lifetime.rs"

[[test]]
name = "setter_custom"
path = "tests/setter_custom.rs"

[[test]]
name = "setter_extend"
path = "tests/setter_extend.rs"

[[test]]
name = "setter_into"
path = "tests/setter_into.rs"

[[test]]
name = "setter_name"
path = "tests/setter_name.rs"

[[test]]
name = "setter_pattern"
path = "tests/setter_pattern.rs"

[[test]]
name = "setter_prefix"
path = "tests/setter_prefix.rs"

[[test]]
name = "setter_strip_option"
path = "tests/setter_strip_option.rs"

[[test]]
name = "setter_visibility"
path = "tests/setter_visibility.rs"

[[test]]
name = "skip-setter"
path = "tests/skip-setter.rs"

[[test]]
name = "try_setter"
path = "tests/try_setter.rs"

[[test]]
name = "validation"
path = "tests/validation.rs"

[dependencies.derive_builder_macro]
version = "=0.20.2"

[dev-dependencies.rustversion]
version = "1.0.4"

[dev-dependencies.serde]
version = "1"
features = ["derive"]

[dev-dependencies.serde_json]
version = "1"

[features]
alloc = ["derive_builder_macro/alloc"]
clippy = ["derive_builder_macro/clippy"]
default = ["std"]
std = ["derive_builder_macro/lib_has_std"]

[target."cfg(compiletests)".dev-dependencies.trybuild]
version = "1.0.38"

[lints.rust.unexpected_cfgs]
level = "warn"
priority = 0
check-cfg = ["cfg(compiletests)"]
