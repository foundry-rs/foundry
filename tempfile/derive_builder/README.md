![Build](https://github.com/colin-kiegel/rust-derive-builder/workflows/Build/badge.svg)
[![Rust version](https://img.shields.io/badge/rust-1.56+-blue.svg)]()
[![Documentation](https://docs.rs/derive_builder/badge.svg)](https://docs.rs/derive_builder)
[![Latest version](https://img.shields.io/crates/v/derive_builder.svg)](https://crates.io/crates/derive_builder)
[![All downloads](https://img.shields.io/crates/d/derive_builder.svg)](https://crates.io/crates/derive_builder)
[![Downloads of latest version](https://img.shields.io/crates/dv/derive_builder.svg)](https://crates.io/crates/derive_builder)

# Builder Pattern Derive

[Rust][rust] macro to automatically implement the **builder pattern** for arbitrary structs. A simple `#[derive(Builder)]` will generate a `FooBuilder` for your struct `Foo` with all setter-methods and a build method.

## How it Works

```rust
use derive_builder::Builder;

#[derive(Default, Builder, Debug)]
#[builder(setter(into))]
struct Channel {
    token: i32,
    special_info: i32,
    // .. a whole bunch of other fields ..
}

fn main() {
    // builder pattern, go, go, go!...
    let ch = ChannelBuilder::default()
        .special_info(42u8)
        .token(19124)
        .build()
        .unwrap();
    println!("{:?}", ch);
}
```

Note that we did not write any definition or implementation of `ChannelBuilder`. Instead the `derive_builder` crate acts on `#[derive(Builder)]` and generates the necessary code at compile time.

This is the generated boilerplate code you didn't need to write. :-)

```rust,ignore
#[derive(Clone, Default)]
struct ChannelBuilder {
    token: Option<i32>,
    special_info: Option<i32>,
}

#[allow(dead_code)]
impl ChannelBuilder {
    pub fn token<VALUE: Into<i32>>(&mut self, value: VALUE) -> &mut Self {
        let mut new = self;
        new.token = Some(value.into());
        new
    }
    pub fn special_info<VALUE: Into<i32>>(&mut self, value: VALUE) -> &mut Self {
        let mut new = self;
        new.special_info = Some(value.into());
        new
    }
    fn build(
        &self,
    ) -> Result<Channel, ChannelBuilderError> {
        Ok(Channel {
            id: match self.id {
                Some(ref value) => Clone::clone(value),
                None => {
                    return Err(
                        Into::into(
                            ::derive_builder::UninitializedFieldError::from("id"),
                        ),
                    )
                }
            },
            token: match self.token {
                Some(ref value) => Clone::clone(value),
                None => {
                    return Err(
                        Into::into(
                            ::derive_builder::UninitializedFieldError::from("token"),
                        ),
                    )
                }
            },
            special_info: match self.special_info {
                Some(ref value) => Clone::clone(value),
                None => {
                    return Err(
                        Into::into(
                            ::derive_builder::UninitializedFieldError::from("special_info"),
                        ),
                    )
                }
            },
        })
    }
}
```

_Note: This is edited for readability. The generated code doesn't assume traits such as `Into` are in-scope, and uses full paths to access them._

## Get Started

It's as simple as three steps:

1. Add `derive_builder` to your `Cargo.toml` either manually or
   with [cargo-edit](https://github.com/killercup/cargo-edit):

-   `cargo add derive_builder`

2. Add `use derive_builder::Builder;`
3. Annotate your struct with `#[derive(Builder)]`

## Usage and Features

-   **Chaining**: The setter calls can be chained, because they consume and return `&mut self` by default.
-   **Builder patterns**: You can opt into other builder patterns by preceding your struct (or field) with `#[builder(pattern = "owned")]` or `#[builder(pattern = "immutable")]`.
-   **Extensible**: You can still define your own implementations for the builder struct and define additional methods. Just make sure to name them differently than the setter and build methods.
-   **Documentation and attributes**: Setter methods can be documented by simply documenting the corresponding field. Similarly `#[cfg(...)]` and `#[allow(...)]` attributes are also applied to the setter methods.
-   **Hidden fields**: You can skip setters via `#[builder(setter(skip))]` on each field individually.
-   **Setter visibility**: You can opt into private setter by preceding your struct with `#[builder(private)]`.
-   **Setter type conversions**: With `#[builder(setter(into))]`, setter methods will be generic over the input types â€“ you can then supply every argument that implements the [`Into`][into] trait for the field type.
-   **Setter strip option**: With `#[builder(setter(strip_option))]`, setter methods will take `T` as parameter'type for field of type `Option<T>`.
-   **Collection setters**: Adding `#[builder(setter(each(name = "method_name")))]` to fields whose types implement `Default` and `Extend` will generate a setter which adds items to the builder collection for that field. It's possible for these setters to be generic over the `Into<T>` trait too, like so: `#[builder(setter(each(name = "foo", into)))]`.
-   **Builder field visibility**: You can use `#[builder(field(private))]` or `..(public)`, to set field visibility of your builder.
-   **Generic structs**: Are also supported, but you **must not** use a type parameter named `VALUE`, if you also activate setter type conversions.
-   **Default values**: You can use `#[builder(default)]` to delegate to the `Default` implementation or any explicit value via ` = ".."`. This works both on the struct and field level.
-   **Pre-build validation**: You can use `#[builder(build_fn(validate = "path::to::fn"))]` to add your own validation before the target struct is generated.
-   **Build method suppression**: You can use `#[builder(build_fn(skip))]` to disable auto-implementation of the build method and provide your own.
-   **Custom build method error types**: You can use `#[builder(build_fn(error = "path::to::Error"))]` to have your builder return an error type of your choosing. By default, the macro will emit an error type alongside the builder.
-   **Builder derivations**: You can use `#[builder(derive(Trait1, Trait2, ...))]` to have the builder derive additonal traits. All builders derive `Default` and `Clone`, so you should not declare those in this attribute.
-   **Pass-through attributes**: Use `#[builder_struct_attr(...)]`, `#[builder_impl_attr(...)]`, `#[builder_field_attr(...)]`, and `#[builder_setter_attr(...)]` to declare attributes that will be added to the relevant part of the generated builder.
-   **no_std support**: Just add `#[builder(no_std)]` to your struct, use feature `alloc`, and add `extern crate alloc` to your crate.
-   **No alloc no_std support**: Do not use `alloc` feature and then either add `#[builder(no_std, build_fn(error(validation_error = false)))]` or `#[builder(no_std, build_fn(error = "path::to::Error"))]` to your struct.
-   **Renaming and re-export support**: Use `#[builder(crate = "...")]` to set the root for `derive_builder`. This is useful if you want to rename `derive_builder` in `Cargo.toml` or if your crate is re-exporting `derive_builder::Builder` and needs the generated code to not directly reference the `derive_builder` crate.

For more information and examples please take a look at our [documentation][doc].

## Gotchas

-   Tuple structs and unit structs are not supported as they have no field names. We do not intend to support them.
-   When defining a generic struct, you cannot use `VALUE` as a generic parameter as this is what all setters are using.

## [Documentation][doc]

Detailed explaination of all features and tips for troubleshooting. You'll also find a discussion of different builder patterns.

[doc]: https://colin-kiegel.github.io/rust-derive-builder
[rust]: https://www.rust-lang.org/
[builder-pattern]: https://aturon.github.io/ownership/builders.html
[into]: https://doc.rust-lang.org/nightly/std/convert/trait.Into.html

## [Changelog](CHANGELOG.md)

Yes, we keep a changelog.

## License

Licensed under either of

-   Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or <http://www.apache.org/licenses/LICENSE-2.0>)
-   MIT license ([LICENSE-MIT](LICENSE-MIT) or <http://opensource.org/licenses/MIT>)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally
submitted for inclusion in the work by you, as defined in the Apache-2.0
license, shall be dual licensed as above, without any additional terms or
conditions.
