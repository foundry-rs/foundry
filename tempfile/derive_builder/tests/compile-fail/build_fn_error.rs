use derive_builder::Builder;

#[derive(Builder)]
#[builder(build_fn(error(validation_error = false, path = "hello")))]
struct Fee {
    hi: u32,
}

#[derive(Builder)]
#[builder(build_fn(error(validation_error = false), validate = "hello"))]
struct Foo {
    hi: u32,
}

#[derive(Builder)]
#[builder(build_fn(error(path = "hello")))]
struct Fii {
    hi: u32,
}

#[derive(Builder)]
#[builder(build_fn(error()))]
struct Fum {
    hi: u32,
}

fn main() {}