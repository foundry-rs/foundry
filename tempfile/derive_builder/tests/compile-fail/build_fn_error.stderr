error: Unknown field: `path`
 --> tests/compile-fail/build_fn_error.rs:4:52
  |
4 | #[builder(build_fn(error(validation_error = false, path = "hello")))]
  |                                                    ^^^^

error: Cannot set `error(validation_error = false)` when using `validate`
  --> tests/compile-fail/build_fn_error.rs:10:45
   |
10 | #[builder(build_fn(error(validation_error = false), validate = "hello"))]
   |                                             ^^^^^

error: Unknown field: `path`
  --> tests/compile-fail/build_fn_error.rs:16:26
   |
16 | #[builder(build_fn(error(path = "hello")))]
   |                          ^^^^

error: Missing field `validation_error` at build_fn/error
  --> tests/compile-fail/build_fn_error.rs:15:10
   |
15 | #[derive(Builder)]
   |          ^^^^^^^
   |
   = note: this error originates in the derive macro `Builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Missing field `validation_error`
  --> tests/compile-fail/build_fn_error.rs:22:20
   |
22 | #[builder(build_fn(error()))]
   |                    ^^^^^
